planet/api/http.py

```diff
@@ -13,135 +13,170 @@
 # the License.
 
 """Functionality to perform HTTP requests"""
-
-# import http.client as http_client
+import asyncio
 import logging
-import os
-import re
-import time
 
-from requests import Session
-from requests.compat import urlparse
+import httpx
 
 from . import exceptions, models
 from . __version__ import __version__
 
-
-LOGGER = logging.getLogger(__name__)
-
-DEFAULT_CHUNK_SIZE = 32 * 1024
-
-USE_STRICT_SSL = not (os.getenv('DISABLE_STRICT_SSL', '').lower() == 'true')
-
+RETRY_COUNT = 5
 RETRY_WAIT_TIME = 1  # seconds
 
+LOGGER = logging.getLogger(__name__)
 
-# def setup_logging():
-#     log_level = LOGGER.getEffectiveLevel()
-#     urllib3_logger = logging.getLogger(
-#         'requests.packages.urllib3')
-#     urllib3_logger.setLevel(log_level)
-#
-#     # if debug level set then its nice to see the headers of the request
-#     if log_level == logging.DEBUG:
-#         http_client.HTTPConnection.set_debuglevel(1)
-#     else:
-#         http_client.HTTPConnection.set_debuglevel(0)
-#
 
-def _log_request(req):
-    LOGGER.info('%s %s %s %s', req.method, req.url, req.params, req.data)
+class APlanetSession():
+    '''Context manager for asynchronous communication with the Planet server.
 
+    Authentication for Planet servers is given as ('', '<api key>').
 
-class PlanetSession():
-    """Provides communication with the Planet server"""
+    :param auth: Planet server authentication.
+    :type auth: httpx.Auth or tuple.
+    '''
 
-    def __init__(self):
-        # general session for sync api calls
-        self._session = RedirectSession()
-        self._session.headers.update({'User-Agent': self._get_user_agent()})
-        self._session.verify = USE_STRICT_SSL
+    def __init__(self, auth=None):
+        self._client = httpx.AsyncClient(auth=auth)
+        self._client.headers.update({'User-Agent': self._get_user_agent()})
+        self._client.event_hooks['request'] = [self._log_request]
+        self._client.event_hooks['response'] = [
+            self._log_response,
+            self._raise_for_status
+        ]
         self.retry_wait_time = RETRY_WAIT_TIME
+        self.retry_count = RETRY_COUNT
 
-    @staticmethod
-    def _get_user_agent():
-        return 'planet-client-python/' + __version__
-
-    def __enter__(self):
+    async def __aenter__(self):
         return self
 
-    def __exit__(self, *args):
-        self.close()
+    async def __aexit__(self, *args):
+        await self.aclose()
 
-    def close(self):
-        self._session.close()
+    async def aclose(self):
+        await self._client.aclose()
 
-    def request(self, request, retry_count=5):
-        '''Submit a request with retry.
+    async def retry(self, func, *a, **kw):
+        '''Run an asynchronous request function with retry.'''
+        retry_count = self.retry_count
+        wait_time = self.retry_wait_time
```

Comment:
instead of a static wait time, this could use exponential backoff

Reply:
ohhhhh yes, I agree fully! Actually, when the next version of httpx is released, retries with exponential backoff will be supported. I had that documented close-by but should also document that here.